### PR TITLE
Bug 1243154 - Improve loading times for logins when querying database

### PR DESCRIPTION
**Pull request by sleroux**

### CODE REVIEW

**Comment by sleroux** [comment]
Output of query plan after indexes were added:

```
sqlite> EXPLAIN QUERY PLAN SELECT guid, username, password, hostname, httpRealm, formSubmitURL, usernameField, passwordField
   ...> FROM loginsL
   ...> WHERE is_deleted = 0
   ...> UNION ALL
   ...> SELECT guid, username, password, hostname, httpRealm, formSubmitURL, usernameField, passwordField
   ...> FROM loginsM
   ...> WHERE is_overridden = 0
   ...> ORDER BY hostname ASC
   ...> ;
1|0|0|SEARCH TABLE loginsL USING INDEX idx_loginsL_is_deleted_hostname (is_deleted=?)
2|0|0|SEARCH TABLE loginsM USING INDEX idx_loginsM_is_overriden_hostname (is_overridden=?)
0|0|0|COMPOUND SUBQUERIES 1 AND 2 (UNION ALL)
```

File: Client/Frontend/Login Management/LoginListViewController.swift [hunk 1]
```diff
@@ -330,8 +330,12 @@ extension LoginListViewController: SearchInputViewDelegate {
     }
 
     private func searchLoginsWithText(text: String) -> Success {
-        activeLoginQuery = profile.logins.searchLoginsWithQuery(text)
-            .bindQueue(dispatch_get_main_queue(), f: reloadTableWithResult)
+        activeLoginQuery =
+            (text.isEmpty ?
+                profile.logins.getAllLogins()
+                    .bindQueue(dispatch_get_main_queue(), f: reloadTableWithResult) :
```
Comment (rnewman) [inline Client/Frontend/Login Management/LoginListViewController.swift]: Gentle preference for phrasing this as:

```
let deferred = text.isEmpty ? profile.logins.getAllLogins() : profile.logins.searchLoginsWithQuery(text)
return deferred.bindQueue(dispatch_get_main_queue(), f: reloadTableWithResult)
```

but why not just make `SQLiteLogins.searchLoginsWithQuery` fall through to `getAllLogins` when passed an empty string?

File: Storage/SQL/SQLiteLogins.swift [hunk 2]
```diff
@@ -10,12 +10,14 @@ private let log = Logger.syncLogger
 
 let TableLoginsMirror = "loginsM"
 let TableLoginsLocal = "loginsL"
+let IndexLoginsOverrideHostname = "idx_loginsM_is_overriden_hostname"
```
Comment (rnewman) [inline Storage/SQL/SQLiteLogins.swift]: `overridden`

File: Storage/SQL/SQLiteLogins.swift [hunk 3]
```diff
@@ -83,6 +84,20 @@ private class LoginsTable: Table {
             return drop(db) && create(db)
         }
 
+        if from < 3 && to >= 3 {
+            let overrideHostnameIndex =
+            "CREATE INDEX IF NOT EXISTS \(IndexLoginsOverrideHostname) " +
+            "ON \(TableLoginsMirror) (is_overridden, hostname)"
```
Comment (rnewman) [inline Storage/SQL/SQLiteLogins.swift]: Put both of these queries on one line. Helps with the compilation speed.

File: Storage/SQL/SQLiteLogins.swift [hunk 4]
```diff
@@ -83,6 +84,20 @@ private class LoginsTable: Table {
...
+            "ON \(TableLoginsLocal) (is_deleted, hostname)"
+
+            if !self.run(db, queries: [overrideHostnameIndex, deletedHostnameIndex]) {
+                return false
+            }
+        }
+
         // TODO: real update!
         log.debug("Updating logins table from \(from) to \(to).")
         return drop(db) && create(db)
```
Comment (rnewman) [inline Storage/SQL/SQLiteLogins.swift]: Nope nope nope nope

**Comment by rnewman** [comment]
You also need to create the indices in `create`, otherwise new users won't get them.

**Comment by sleroux** [comment]
@rnewman I've updated the PR to include the DB bug fixes and searchWithQuery suggestion.

File: Client/Frontend/Login Management/LoginListViewController.swift [hunk 5]
```diff
@@ -330,8 +330,7 @@ extension LoginListViewController: SearchInputViewDelegate {
     }
 
     private func searchLoginsWithText(text: String) -> Success {
-        activeLoginQuery = profile.logins.searchLoginsWithQuery(text)
-            .bindQueue(dispatch_get_main_queue(), f: reloadTableWithResult)
+        activeLoginQuery = profile.logins.searchLoginsWithQuery(text).bindQueue(dispatch_get_main_queue(), f: reloadTableWithResult)
```
Comment (rnewman) [inline Client/Frontend/Login Management/LoginListViewController.swift]: You can probably drop this linebreak change altogether, save the blame.

File: Storage/SQL/SQLiteLogins.swift [hunk 6]
```diff
@@ -10,12 +10,19 @@ private let log = Logger.syncLogger
 
 let TableLoginsMirror = "loginsM"
 let TableLoginsLocal = "loginsL"
+let IndexLoginsOverrideHostname = "idx_loginsM_is_overridden_hostname"
+let IndexLoginsDeletedHostname = "idx_loginsL_is_deleted_hostname"
 let AllLoginTables: [String] = [TableLoginsMirror, TableLoginsLocal]
 
+private let OverridenHostnameIndexQuery =
```
Comment (rnewman) [inline Storage/SQL/SQLiteLogins.swift]: s/Overriden/Overridden/g

**Comment by rnewman** [comment]
LGTM with those two nits.

**Comment by sleroux** [comment]
Thanks! I've folded the commits and will land.

**Pull request merged**